File: sdk/go/target.go
package sykli

import (
	"context"
	"fmt"
	"regexp"
	"strings"
)

// =============================================================================
// TARGET INTERFACE - The Core
// =============================================================================

// Target is the core interface - just one method.
//
// A Target is something that can run tasks. That's it.
//
// Simple targets (GitHub Actions, SSH, Lambda) implement only this.
// Complex targets opt into additional capabilities.
//
// Example - Simple target:
//
//	type GitHubActionsTarget struct{}
//
//	func (g *GitHubActionsTarget) RunTask(ctx context.Context, task TaskSpec) Result {
//	    // Trigger workflow, wait for completion
//	    return Result{Success: true}
//	}
//
// Example - With capabilities:
//
//	type MyTarget struct {
//	    EnvSecrets   // Embed for environment-based secrets
//	    LocalStorage // Embed for filesystem storage
//	}
//
//	func (m *MyTarget) RunTask(ctx context.Context, task TaskSpec) Result {
//	    // Use embedded capabilities + custom logic
//	}
type Target interface {
	// RunTask executes a single task.
	//
	// This is the ONLY required method. Everything else is optional.
	RunTask(ctx context.Context, task TaskSpec) Result
}

// =============================================================================
// OPTIONAL CAPABILITIES - Implement what you need
// =============================================================================

// Lifecycle adds setup/teardown around pipeline execution.
//
// Implement this if your target needs to:
//   - Initialize connections before running
//   - Clean up after the pipeline completes
//   - Maintain state across tasks
//
// Example:
//
//	type MyTarget struct {
//	    conn *Connection
//	}
//
//	func (m *MyTarget) Setup(ctx context.Context) error {
//	    m.conn = connect()
//	    return nil
//	}
//
//	func (m *MyTarget) Teardown(ctx context.Context) error {
//	    return m.conn.Close()
//	}
type Lifecycle interface {
	Setup(ctx context.Context) error
	Teardown(ctx context.Context) error
}

// Secrets provides secret resolution.
//
// Implement this if your target can provide secret values
// (API keys, tokens, passwords) to tasks.
//
// Example:
//
//	func (m *MyTarget) ResolveSecret(ctx context.Context, name string) (string, error) {
//	    return vault.Read("secret/" + name)
//	}
type Secrets interface {
	ResolveSecret(ctx context.Context, name string) (string, error)
}

// Storage provides volume and artifact management.
//
// Implement this if your target needs to:
//   - Create storage volumes for tasks
//   - Pass artifacts between tasks
//   - Persist build outputs
type Storage interface {
	CreateVolume(ctx context.Context, name string, opts VolumeOptions) (Volume, error)
	ArtifactPath(taskName, artifactName string) string
	CopyArtifact(ctx context.Context, src, dst string) error
}

// Services provides service container management.
//
// Implement this if your target can run background services
// (databases, caches) that tasks can connect to.
type Services interface {
	StartServices(ctx context.Context, taskName string, services []ServiceSpec) (interface{}, error)
	StopServices(ctx context.Context, networkInfo interface{}) error
}

// =============================================================================
// CAPABILITY CHECKING
// =============================================================================

// HasLifecycle checks if a target implements Lifecycle.
func HasLifecycle(t Target) bool {
	_, ok := t.(Lifecycle)
	return ok
}

// HasSecrets checks if a target implements Secrets.
func HasSecrets(t Target) bool {
	_, ok := t.(Secrets)
	return ok
}

// HasStorage checks if a target implements Storage.
func HasStorage(t Target) bool {
	_, ok := t.(Storage)
	return ok
}

// HasServices checks if a target implements Services.
func HasServices(t Target) bool {
	_, ok := t.(Services)
	return ok
}

// AsLifecycle returns the Lifecycle interface if implemented.
func AsLifecycle(t Target) (Lifecycle, bool) {
	l, ok := t.(Lifecycle)
	return l, ok
}

// AsSecrets returns the Secrets interface if implemented.
func AsSecrets(t Target) (Secrets, bool) {
	s, ok := t.(Secrets)
	return s, ok
}

// AsStorage returns the Storage interface if implemented.
func AsStorage(t Target) (Storage, bool) {
	s, ok := t.(Storage)
	return s, ok
}

// AsServices returns the Services interface if implemented.
func AsServices(t Target) (Services, bool) {
	s, ok := t.(Services)
	return s, ok
}

// =============================================================================
// BUILT-IN CAPABILITY IMPLEMENTATIONS
// =============================================================================

// EnvSecrets resolves secrets from environment variables.
// Embed this in your target to get env-based secrets for free.
//
// Example:
//
//	type MyTarget struct {
//	    EnvSecrets
//	}
type EnvSecrets struct{}

// ResolveSecret reads from environment variables.
func (EnvSecrets) ResolveSecret(ctx context.Context, name string) (string, error) {
	// Implementation in executor - this is just the interface
	return "", nil
}

// =============================================================================
// TYPES
// =============================================================================

// TaskSpec contains all information needed to execute a task.
type TaskSpec struct {
	Name      string
	Command   string
	Image     string            // Container image (empty = shell)
	Workdir   string            // Working directory inside container
	Env       map[string]string // Environment variables
	Mounts    []MountSpec       // Volume mounts
	Timeout   int               // Timeout in seconds (0 = default)
	DependsOn []string          // For informational purposes
	Services  []ServiceSpec     // Service containers for this task

	// Target-specific options (nil if not using that target)
	K8s *K8sTaskOptions
}

// MountSpec describes a volume mount.
type MountSpec struct {
	Volume Volume // The volume to mount
	Path   string // Mount path inside container
}

// ServiceSpec describes a service container.
type ServiceSpec struct {
	Name  string // Service name (used as hostname)
	Image string // Container image
}

// Result contains execution results.
type Result struct {
	Success  bool
	ExitCode int
	Output   string // Captured stdout/stderr
	Duration int64  // Execution time in milliseconds
	Error    error  // Non-nil if execution failed
}

// VolumeOptions for creating volumes.
type VolumeOptions struct {
	Size string // e.g., "1Gi", "10Gi"
}

// Volume represents provisioned storage.
type Volume interface {
	ID() string
	HostPath() string   // Empty for non-local targets
	Reference() string  // Target-specific reference
}

// =============================================================================
// K8S TARGET OPTIONS
// =============================================================================

// K8sOptions provides Kubernetes-specific configuration for a task.
// This is a minimal API covering 95% of CI use cases.
//
// For advanced options (tolerations, affinity, security contexts),
// use K8sRaw() to pass raw JSON.
//
// Example:
//
//	s.Task("build").
//	    Run("go build").
//	    K8s(sykli.K8sOptions{Memory: "4Gi", CPU: "2"})
//
//	s.Task("train").
//	    Run("python train.py").
//	    K8s(sykli.K8sOptions{Memory: "32Gi", GPU: 1})
type K8sOptions struct {
	// Memory sets both request and limit (e.g., "4Gi", "512Mi").
	Memory string

	// CPU sets both request and limit (e.g., "2", "500m").
	CPU string

	// GPU requests NVIDIA GPUs (e.g., 1, 2).
	GPU int
}

// K8sTaskOptions is an alias for backward compatibility.
// Deprecated: Use K8sOptions instead.
type K8sTaskOptions = K8sOptions

// =============================================================================
// TASK K8S EXTENSION
// =============================================================================

// K8s adds Kubernetes-specific options to a task.
//
// Example:
//
//	s.Task("build").Run("go build").K8s(sykli.K8sOptions{Memory: "4Gi", CPU: "2"})
func (t *Task) K8s(opts K8sOptions) *Task {
	t.k8sOptions = &opts
	return t
}

// K8sRaw adds raw Kubernetes configuration as JSON.
// Use this for advanced options not covered by K8sOptions (tolerations, affinity, etc.).
//
// The JSON is passed directly to the executor and merged with K8sOptions.
// K8sOptions fields take precedence over K8sRaw for overlapping settings.
//
// Example:
//
//	// GPU node with toleration
//	s.Task("train").
//	    K8s(sykli.K8sOptions{Memory: "32Gi", GPU: 1}).
//	    K8sRaw(`{"nodeSelector": {"gpu": "true"}, "tolerations": [{"key": "gpu", "effect": "NoSchedule"}]}`)
func (t *Task) K8sRaw(jsonConfig string) *Task {
	t.k8sRaw = jsonConfig
	return t
}

// MergeK8sOptions merges defaults with task-specific options.
// Task options override defaults.
func MergeK8sOptions(defaults, task *K8sOptions) *K8sOptions {
	if defaults == nil {
		return task
	}
	if task == nil {
		copy := *defaults
		return &copy
	}

	result := *defaults

	if task.Memory != "" {
		result.Memory = task.Memory
	}
	if task.CPU != "" {
		result.CPU = task.CPU
	}
	if task.GPU > 0 {
		result.GPU = task.GPU
	}

	return &result
}

// =============================================================================
// K8S VALIDATION
// =============================================================================

// K8s quantity patterns - shared validation logic
var (
	// Memory: Ki, Mi, Gi, Ti, Pi, Ei (binary) or k, M, G, T, P, E (decimal)
	k8sMemoryPattern = regexp.MustCompile(`^[0-9]+(\.[0-9]+)?(Ki|Mi|Gi|Ti|Pi|Ei|k|M|G|T|P|E)?$`)
	// CPU: whole numbers, decimals, or millicores (e.g., "100m", "0.5", "2")
	k8sCPUPattern = regexp.MustCompile(`^[0-9]+(\.[0-9]+)?m?$`)
)

// K8sValidationError contains details about K8s option validation failures.
type K8sValidationError struct {
	Field   string
	Value   string
	Message string
}

func (e K8sValidationError) Error() string {
	return fmt.Sprintf("k8s.%s: %s (got %q)", e.Field, e.Message, e.Value)
}

// ValidateK8sOptions validates K8s options and returns all errors found.
// Returns nil if validation passes.
func ValidateK8sOptions(opts *K8sOptions) []error {
	if opts == nil {
		return nil
	}

	var errs []error

	// Validate memory
	if opts.Memory != "" {
		if err := validateK8sMemory("memory", opts.Memory); err != nil {
			errs = append(errs, err)
		}
	}

	// Validate CPU
	if opts.CPU != "" {
		if err := validateK8sCPU("cpu", opts.CPU); err != nil {
			errs = append(errs, err)
		}
	}

	return errs
}

func validateK8sMemory(field, value string) error {
	if value == "" {
		return nil
	}
	if !k8sMemoryPattern.MatchString(value) {
		suggestion := ""
		lower := strings.ToLower(value)
		if strings.HasSuffix(lower, "gb") {
			suggestion = " (did you mean 'Gi'?)"
		} else if strings.HasSuffix(lower, "mb") {
			suggestion = " (did you mean 'Mi'?)"
		} else if strings.HasSuffix(lower, "kb") {
			suggestion = " (did you mean 'Ki'?)"
		}
		return K8sValidationError{
			Field:   field,
			Value:   value,
			Message: "invalid memory format, use Ki/Mi/Gi/Ti (e.g., '512Mi', '4Gi')" + suggestion,
		}
	}
	return nil
}

func validateK8sCPU(field, value string) error {
	if value == "" {
		return nil
	}
	if !k8sCPUPattern.MatchString(value) {
		return K8sValidationError{
			Field:   field,
			Value:   value,
			Message: "invalid CPU format, use cores or millicores (e.g., '500m', '0.5', '2')",
		}
	}
	return nil
}

File: sdk/rust/src/target.rs
//! Target interface - where pipelines execute.
//!
//! A Target is simply something that can run tasks. That's it.
//!
//! # The Minimal Interface
//!
//! ```rust,ignore
//! use sykli::target::{Target, TaskSpec, Result};
//!
//! struct MyTarget;
//!
//! impl Target for MyTarget {
//!     fn run_task(&self, task: &TaskSpec) -> Result {
//!         // Execute the task somehow
//!         Result::success()
//!     }
//! }
//! ```
//!
//! That's all you need. One method.
//!
//! # Optional Capabilities
//!
//! Targets can opt into additional capabilities by implementing
//! optional traits. These are building blocks - use what you need:
//!
//! - [`Lifecycle`] - setup/teardown around pipeline execution
//! - [`Secrets`] - resolve secrets by name
//! - [`Storage`] - manage volumes and artifacts
//! - [`Services`] - start/stop service containers
//!
//! # Examples
//!
//! ## Simple Target (GitHub Actions)
//!
//! ```rust,ignore
//! struct GitHubActionsTarget {
//!     workflow: String,
//! }
//!
//! impl Target for GitHubActionsTarget {
//!     fn run_task(&self, task: &TaskSpec) -> Result {
//!         // Trigger workflow via GitHub API
//!         // Wait for completion
//!         // Return result
//!         Result::success()
//!     }
//! }
//! ```
//!
//! ## Full-Featured Target
//!
//! ```rust,ignore
//! struct MyK8sTarget {
//!     namespace: String,
//!     kubeconfig: Option<String>,
//! }
//!
//! impl Target for MyK8sTarget {
//!     fn run_task(&self, task: &TaskSpec) -> Result { ... }
//! }
//!
//! impl Lifecycle for MyK8sTarget {
//!     fn setup(&mut self) -> std::result::Result<(), Error> { ... }
//!     fn teardown(&mut self) -> std::result::Result<(), Error> { ... }
//! }
//!
//! impl Secrets for MyK8sTarget {
//!     fn resolve_secret(&self, name: &str) -> std::result::Result<String, Error> { ... }
//! }
//! ```

use std::collections::HashMap;
use std::error::Error as StdError;
use std::fmt;
use std::time::Duration;

// =============================================================================
// ERROR TYPE
// =============================================================================

/// Error type for target operations.
#[derive(Debug)]
pub struct Error {
    message: String,
    source: Option<Box<dyn StdError + Send + Sync>>,
}

impl Error {
    /// Creates a new error with the given message.
    pub fn new(message: impl Into<String>) -> Self {
        Self {
            message: message.into(),
            source: None,
        }
    }

    /// Creates an error with a source cause.
    pub fn with_source(message: impl Into<String>, source: impl StdError + Send + Sync + 'static) -> Self {
        Self {
            message: message.into(),
            source: Some(Box::new(source)),
        }
    }
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.message)
    }
}

impl StdError for Error {
    fn source(&self) -> Option<&(dyn StdError + 'static)> {
        self.source.as_ref().map(|e| e.as_ref() as _)
    }
}

// =============================================================================
// TASK SPEC
// =============================================================================

/// Task specification passed to [`Target::run_task`].
#[derive(Debug, Clone)]
pub struct TaskSpec {
    /// Task name.
    pub name: String,
    /// Command to execute.
    pub command: String,
    /// Container image (empty = shell execution).
    pub image: Option<String>,
    /// Working directory inside container.
    pub workdir: Option<String>,
    /// Environment variables.
    pub env: HashMap<String, String>,
    /// Volume mounts.
    pub mounts: Vec<MountSpec>,
    /// Timeout in seconds.
    pub timeout: Option<u32>,
    /// Service containers for this task.
    pub services: Vec<ServiceSpec>,
}

impl TaskSpec {
    /// Creates a new task spec with the given name and command.
    pub fn new(name: impl Into<String>, command: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            command: command.into(),
            image: None,
            workdir: None,
            env: HashMap::new(),
            mounts: Vec::new(),
            timeout: None,
            services: Vec::new(),
        }
    }
}

/// Volume mount specification.
#[derive(Debug, Clone)]
pub struct MountSpec {
    /// Host path or volume reference.
    pub source: String,
    /// Mount path inside container.
    pub target: String,
    /// Mount type (directory, cache).
    pub mount_type: MountType,
}

/// Type of mount.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MountType {
    /// Directory mount (bind mount).
    Directory,
    /// Named cache volume.
    Cache,
}

/// Service container specification.
#[derive(Debug, Clone)]
pub struct ServiceSpec {
    /// Service name (used as hostname).
    pub name: String,
    /// Container image.
    pub image: String,
}

// =============================================================================
// RESULT
// =============================================================================

/// Result from task execution.
#[derive(Debug, Clone)]
pub struct Result {
    /// Whether the task succeeded.
    pub success: bool,
    /// Exit code.
    pub exit_code: i32,
    /// Captured output (stdout/stderr).
    pub output: String,
    /// Execution duration.
    pub duration: Duration,
    /// Error message if failed.
    pub error: Option<String>,
}

impl Result {
    /// Creates a successful result.
    pub fn success() -> Self {
        Self {
            success: true,
            exit_code: 0,
            output: String::new(),
            duration: Duration::ZERO,
            error: None,
        }
    }

    /// Creates a successful result with output.
    pub fn success_with_output(output: impl Into<String>, duration: Duration) -> Self {
        Self {
            success: true,
            exit_code: 0,
            output: output.into(),
            duration,
            error: None,
        }
    }

    /// Creates a failed result.
    pub fn failure(exit_code: i32, output: impl Into<String>) -> Self {
        Self {
            success: false,
            exit_code,
            output: output.into(),
            duration: Duration::ZERO,
            error: None,
        }
    }

    /// Creates a failed result with an error message.
    pub fn error(message: impl Into<String>) -> Self {
        Self {
            success: false,
            exit_code: 1,
            output: String::new(),
            duration: Duration::ZERO,
            error: Some(message.into()),
        }
    }
}

// =============================================================================
// THE CORE TRAIT - Just one method
// =============================================================================

/// The core Target trait - just one method.
///
/// A Target is something that can run tasks. That's it.
///
/// # Example
///
/// ```rust,ignore
/// use sykli::target::{Target, TaskSpec, Result};
///
/// struct ShellTarget;
///
/// impl Target for ShellTarget {
///     fn run_task(&self, task: &TaskSpec) -> Result {
///         // Execute via shell
///         let output = std::process::Command::new("sh")
///             .arg("-c")
///             .arg(&task.command)
///             .output()
///             .expect("failed to execute");
///
///         if output.status.success() {
///             Result::success_with_output(
///                 String::from_utf8_lossy(&output.stdout),
///                 std::time::Duration::ZERO,
///             )
///         } else {
///             Result::failure(
///                 output.status.code().unwrap_or(1),
///                 String::from_utf8_lossy(&output.stderr),
///             )
///         }
///     }
/// }
/// ```
pub trait Target {
    /// Execute a task.
    ///
    /// This is the ONLY required method. Everything else is optional.
    fn run_task(&self, task: &TaskSpec) -> Result;
}

// =============================================================================
// OPTIONAL CAPABILITIES
// =============================================================================

/// Optional capability: Setup and teardown around pipeline execution.
///
/// Implement this if your target needs to:
/// - Initialize connections before running
/// - Clean up after the pipeline completes
/// - Maintain state across tasks
///
/// # Example
///
/// ```rust,ignore
/// struct DatabaseTarget {
///     connection: Option<Connection>,
/// }
///
/// impl Lifecycle for DatabaseTarget {
///     fn setup(&mut self) -> std::result::Result<(), Error> {
///         self.connection = Some(Connection::new()?);
///         Ok(())
///     }
///
///     fn teardown(&mut self) -> std::result::Result<(), Error> {
///         if let Some(conn) = self.connection.take() {
///             conn.close()?;
///         }
///         Ok(())
///     }
/// }
/// ```
pub trait Lifecycle {
    /// Initialize the target before pipeline execution.
    fn setup(&mut self) -> std::result::Result<(), Error>;

    /// Clean up after pipeline execution.
    fn teardown(&mut self) -> std::result::Result<(), Error>;
}

/// Optional capability: Resolve secrets by name.
///
/// Implement this if your target can provide secret values
/// (API keys, tokens, passwords) to tasks.
///
/// # Example
///
/// ```rust,ignore
/// impl Secrets for VaultTarget {
///     fn resolve_secret(&self, name: &str) -> std::result::Result<String, Error> {
///         self.vault_client.read(&format!("secret/{}", name))
///             .map_err(|e| Error::with_source("vault read failed", e))
///     }
/// }
/// ```
pub trait Secrets {
    /// Resolve a secret value by name.
    fn resolve_secret(&self, name: &str) -> std::result::Result<String, Error>;
}

/// Volume reference returned by [`Storage::create_volume`].
#[derive(Debug, Clone)]
pub struct Volume {
    /// Unique volume ID.
    pub id: String,
    /// Host path (if applicable).
    pub host_path: Option<String>,
    /// Target-specific reference.
    pub reference: String,
}

/// Options for volume creation.
#[derive(Debug, Clone, Default)]
pub struct VolumeOptions {
    /// Volume size (e.g., "1Gi").
    pub size: Option<String>,
}

/// Optional capability: Manage volumes and artifacts.
///
/// Implement this if your target needs to:
/// - Create storage volumes for tasks
/// - Pass artifacts between tasks
/// - Persist build outputs
pub trait Storage {
    /// Create a storage volume.
    fn create_volume(&self, name: &str, opts: &VolumeOptions) -> std::result::Result<Volume, Error>;

    /// Get the path where an artifact should be stored.
    fn artifact_path(&self, task_name: &str, artifact_name: &str) -> String;

    /// Copy an artifact from source to destination.
    fn copy_artifact(&self, src: &str, dst: &str) -> std::result::Result<(), Error>;
}

/// Network info returned by [`Services::start_services`].
#[derive(Debug, Clone)]
pub struct NetworkInfo {
    /// Network name or ID.
    pub network: String,
    /// Container IDs.
    pub containers: Vec<String>,
}

/// Optional capability: Run service containers.
///
/// Implement this if your target can run background services
/// (databases, caches) that tasks can connect to.
pub trait Services {
    /// Start service containers for a task.
    fn start_services(&self, task_name: &str, services: &[ServiceSpec]) -> std::result::Result<NetworkInfo, Error>;

    /// Stop and clean up service containers.
    fn stop_services(&self, network_info: &NetworkInfo) -> std::result::Result<(), Error>;
}

// =============================================================================
// CAPABILITY CHECKING
// =============================================================================

/// Check if a target implements [`Lifecycle`].
pub fn has_lifecycle<T: Target + ?Sized>(_target: &T) -> bool {
    // In Rust, we can't do runtime trait checking like Go/Elixir
    // Users should use the Any trait or explicit type checks
    false
}

// =============================================================================
// BUILT-IN IMPLEMENTATIONS
// =============================================================================

/// Resolves secrets from environment variables.
///
/// # Example
///
/// ```rust,ignore
/// struct MyTarget {
///     secrets: EnvSecrets,
/// }
///
/// // Delegate to EnvSecrets
/// impl Secrets for MyTarget {
///     fn resolve_secret(&self, name: &str) -> std::result::Result<String, Error> {
///         self.secrets.resolve_secret(name)
///     }
/// }
/// ```
#[derive(Debug, Clone, Default)]
pub struct EnvSecrets;

impl Secrets for EnvSecrets {
    fn resolve_secret(&self, name: &str) -> std::result::Result<String, Error> {
        std::env::var(name).map_err(|_| Error::new(format!("secret not found: {}", name)))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    struct TestTarget;

    impl Target for TestTarget {
        fn run_task(&self, _task: &TaskSpec) -> Result {
            Result::success()
        }
    }

    #[test]
    fn test_target_trait() {
        let target = TestTarget;
        let task = TaskSpec::new("test", "echo hello");
        let result = target.run_task(&task);
        assert!(result.success);
    }

    #[test]
    fn test_env_secrets() {
        std::env::set_var("TEST_SECRET", "secret_value");
        let secrets = EnvSecrets;
        let value = secrets.resolve_secret("TEST_SECRET").unwrap();
        assert_eq!(value, "secret_value");
        std::env::remove_var("TEST_SECRET");
    }

    #[test]
    fn test_env_secrets_not_found() {
        let secrets = EnvSecrets;
        let result = secrets.resolve_secret("NONEXISTENT_SECRET");
        assert!(result.is_err());
    }

    #[test]
    fn test_result_success() {
        let result = Result::success();
        assert!(result.success);
        assert_eq!(result.exit_code, 0);
    }

    #[test]
    fn test_result_failure() {
        let result = Result::failure(1, "error output");
        assert!(!result.success);
        assert_eq!(result.exit_code, 1);
        assert_eq!(result.output, "error output");
    }
}
